### PR TITLE
FEATURE: Add word count and indicator when exceeded max

### PR DESCRIPTION
**Pull request by keegangeorge**
**This PR creates a new core reusable component wraps a character counter around any input.**

The component accepts the arguments: `max` (the maximum character limit), `value` (the value of text to be monitored).

It can be used for example, like so:
```hbs
  <CharCounter @max="50" @value={{this.charCounterContent}}>
    <textarea
      placeholder={{i18n "styleguide.sections.char_counter.placeholder"}}
      {{on "input" (action (mut this.charCounterContent) value="target.value")}}
      class="styleguide--char-counter"></textarea>
  </CharCounter>
```

**This PR also:**
1. Applies this component to the chat plugins edit channel's *Edit Description** modal, thereby replacing the simple text area which provided no visual indication when text exceeded the max allowed characters.
2. Adds an example to the `/styleguide` route

Below are some screenshots demonstrating its usage:
| Within adequate word range | Exceeded Range |
|---|---|
|image|image|

### CODE REVIEW

File: plugins/chat/assets/javascripts/discourse/controllers/chat-channel-edit-description.js [hunk 1]
```diff
@@ -2,20 +2,29 @@ import Controller from "@ember/controller";
...
 import ModalFunctionality from "discourse/mixins/modal-functionality";
 import ChatApi from "discourse/plugins/chat/discourse/lib/chat-api";
+import { tracked } from "@glimmer/tracking";
 
 export default class ChatChannelEditDescriptionController extends Controller.extend(
   ModalFunctionality
 ) {
-  editedDescription = "";
+  @tracked editedDescription = "";
+  DESCRIPTION_MAX_LENGTH = 280;
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/controllers/chat-channel-edit-description.js:11]: I prefer to have constants outside of the class

Comment (keegangeorge) [reply plugins/chat/assets/javascripts/discourse/controllers/chat-channel-edit-description.js]: Ah I like the idea of that. Just pushed an update :)

File: plugins/chat/assets/javascripts/discourse/controllers/chat-channel-edit-description.js [hunk 2]
```diff
@@ -2,20 +2,29 @@ import Controller from "@ember/controller";
...
     );
   }
 
+  get wordCount() {
+    return {
+      exceeded: this.editedDescription.length > this.DESCRIPTION_MAX_LENGTH,
+      label: `${this.editedDescription.length}/${this.DESCRIPTION_MAX_LENGTH}`,
+    };
+  }
+
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/controllers/chat-channel-edit-description.js:28]: I would make all of this in the template:

```
get descriptionMaxLength() {
  return DESCRIPTION_MAX_LENGTH;
}
```

```
<DModalBody @title="chat.channel_edit_description_modal.title" @class={{if (gt this.editedDescription.length this.descriptionMaxLength) "exceeded-word-count"}}>
...

{{#if this.editedDescription.length}}
  <small class="chat-channel-edit-description-modal__description-word-count">
    {{this.editedDescription.length}}/{{this.descriptionMaxlength}}
  </small>
{{/if}}
```

**Comment by jjaffeux** [comment]
Generally speaking I think you should encapsulate this in a component (glimmer for extra bonus points), this will also make writing a test super easy.

**Comment by keegangeorge** [comment]
Hey @jjaffeux, I've created the text area and word counter as a single reusable component. I've updated the edit description modal to use it, and I've also implemented it inside the Edit Title modal since it is essentially the same with a different max value. Please let me know what you think.

**Comment by jjaffeux** [comment]
> Hey @jjaffeux, I've created the text area and word counter as a single reusable component. I've updated the edit description modal to use it, and I've also implemented it inside the Edit Title modal since it is essentially the same with a different max value. Please let me know what you think.

Great, thanks! I will look more tomorrow but I have a suggestion:

```
<CharCounter @max={{200}} as |counter|>
  <textarea {{on "input" (action counter.didChange value="target.value")}}>{{@description}}</textarea>
</ChartCounter>
```

That's a quickly thought idea, but I think it's worth exploring.

File: plugins/chat/assets/javascripts/discourse/components/counted-text-area.hbs [hunk 3]
```diff
@@ -0,0 +1,6 @@
+<div ...attributes class={{concat-class "counted-text-area" (if (gt @value.length @max) "exceeded")}}>
+  <textarea {{on "input" (action @onChangeTextAreaValue value="target.value")}}>{{@value}}</textarea>
+  <small>
```
Comment (jjaffeux) [inline plugins/chat/assets/javascripts/discourse/components/counted-text-area.hbs:3]: Isn't it missing some aria helper to make it clear it's changing?

**Comment by keegangeorge** [comment]
Hey @jjaffeux, I've converted it to a block component instead like you suggested. I still have yet to add the `aria` attribute for accessibility but I was wondering what you're thoughts on changing it up a little bit.

I found out we have a `maxLength` attribute for `<textarea>` so we can completely restrict text going beyond the maximum value. So rather than letting the user type beyond the max and see a red border and then them having to backspace, what if we do something like this instead:

image

where the user can see how many characters they have left to type and are prevented from typing beyond the max at all.

Thoughts? cc: @chapoi

**Comment by chapoi** [comment]
> Hey @jjaffeux, I've converted it to a block component instead like you suggested. I still have yet to add the `aria` attribute for accessibility but I was wondering what you're thoughts on changing it up a little bit.
> 
> I found out we have a `maxLength` attribute for `<textarea>` so we can completely restrict text going beyond the maximum value. So rather than letting the user type beyond the max and see a red border and then them having to backspace, what if we do something like this instead:
> 
> image
> 
> where the user can see how many characters they have left to type and are prevented from typing beyond the max at all.
> 
> Thoughts? cc: @chapoi

I rather have it go red. It's much more clear that something is wrong. I also think it's nicer to be able to "finish" what you were writing, and then going to figure out how you're going to rewrite/shorten the whole text.

**Comment by jjaffeux** [comment]
At minimum the <CharCounter> component needs tests

**Comment by keegangeorge** [comment]
@jjaffeux Not sure why tests are failing here. Actual and expected values look the same, and the tests are not failing locally for me. Any ideas?

**Comment by jjaffeux** [comment]
> @jjaffeux Not sure why tests are failing here. Actual and expected values look the same, and the tests are not failing locally for me. Any ideas?

theres a space most likely

**Comment by jjaffeux** [comment]
@keegangeorge I'm sorry, still on my radar ofc, but I have threads and the service objects refactor in high priority atm. Will try to have this sorted next week.

**Pull request merged**